src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -0,0 +1,388 @@
+        /// <exception cref="NotSupportedException">Setting a data section is not supported because the <see cref="EntryType"/> is not <see cref="TarEntryType.RegularFile"/> (or <see cref="TarEntryType.V7RegularFile"/> for an archive of <see cref="TarFormat.V7"/> format).</exception>
+        /// <exception cref="IOException"><para>Cannot set an unreadable stream.</para>
+        /// <para>-or-</para>
+        /// <para>An I/O problem occurred.</para></exception>
+        public Stream? DataStream
+        {
+            get => _header._dataStream;
+            set
+            {
+                if (!IsDataStreamSetterSupported())
+                {
+                    throw new NotSupportedException(SR.TarEntryNotARegularFile);
+                }
+
+                if (value != null && !value.CanRead)
+                {
+                    throw new IOException(SR.IO_NotSupported_UnreadableStream);
+                }
+
+                if (_readerOfOrigin != null)
+                {
+                    // This entry came from a reader, so if the underlying stream is unseekable, we need to
+                    // manually advance the stream pointer to the next header before doing the substitution
+                    // The original stream will get disposed when the reader gets disposed.
+                    _readerOfOrigin.AdvanceDataStreamIfNeeded();
+                    // We only do this once
+                    _readerOfOrigin = null;
+                }
+
+                if (_header._dataStream != null)
+                {
+                    _header._dataStream.Dispose();
+                }
+
+                _header._dataStream = value;
+            }
+        }
+
+        /// <summary>
+        /// A string that represents the current entry.
+        /// </summary>
+        /// <returns>The <see cref="Name"/> of the current entry.</returns>
+        public override string ToString() => Name;
+
+        // Abstract method that determines if setting the data stream for this entry is allowed.
+        internal abstract bool IsDataStreamSetterSupported();
+
+        // Throws an exception if a file exists in the location of 'destinationFileName' and 'overwrite' is 'false',
+        // or if a directory exists in the location of 'destinationFileName'.
+        private static void VerifyOverwriteFileIsPossible(string destinationFileName, bool overwrite)
+        {
+            if (File.Exists(destinationFileName))
+            {
+                if (!overwrite)
+                {
+                    throw new IOException(string.Format(SR.IO_AlreadyExists_Name, destinationFileName));
+                }
+            }
+            // We never want to overwrite a directory, so we always throw
+            else if (Directory.Exists(destinationFileName))
+            {
+                throw new IOException(string.Format(SR.IO_AlreadyExists_Name, destinationFileName));
+            }
```

Comment:
Can we use the new Path.Exists to avoid multiple Exists calls in the non-overwrite case?

Reply:
I tried, but I don't think it would solve this problem: We want to extract a file to the specified location in `destinationFileName`, but we don't know if that destination exists or not, and if it exists, we don't know if it's a file or a directory.

If we find out it exists, and it's a file, we can continue if we specified `overwrite = true`. But if it's a directory, we don't ever want to overwrite it, so we should always throw.

With Path, I don't have a way to differentiate between these two options. I still need to know if the destination is a directory or not.

Reply:
I'm not suggesting instead of. I'm suggesting in addition to. Presumably the vast majority case is it doesn't exist? In which case with Path.Exists we make a single syscall. Only if it does exist would we then use the individual File/Directory methods to throw the right exception.

Reply:
Ah, I see. Thanks for the clarification. I can do that.

Reply:
Fixed.